[MERGE] mail, various: improve activity view, allow to keep done activities

pyduf opened the pull request:
[IMP] {test_}mail: allow to keep done activities and improve activity view

We add the option to keep the completed activities to allow the user to keep
track of the past action, especially the uploaded files which are displayed
along the completed activities. This feature can be activated per activity
type and is disabled by default.

We also improve the activity view to give a better overview of what has been
done and what is still to be done and by whom.

We also add tests for those new features:
- check activity data for the views (server-side test)

* Details about the activity view improvements:

** Improvement of the column header:

- progressbar: previously, only visible activities were counted in the
progression bar. Now all are counted.
- activity counter: instead of displaying the total count (next to the
progression bar), we display now a ratio as 2 numbers (only if keep done is
enabled):

** Improvement of the activity cells:

- add assignees in activity cells of activity view (limit to 2 having the
activity with closest deadline, if more a + is displayed). Note: to avoid
spacing between avatar, we add the optional props noSpacing to the avatar
component.
- add done activity in activity cell popover
- increase by 50% the popover height
- display done activity cell in grey
- instead of displaying the activity count if greater than 1, we display a
ratio ongoing / activity count. Note: we don't display the denominator if all
activities are done or all are ongoing.

** Progress bar filter

The activity filter is slighly modified. Previously, when clicking on a color
of the progress bar, all cell with a different color where hidden (ex.: if we
were clicking on the green area to filter planned activities, red cell which
represents overdue activities were hidden even if there was also a planned
activities for this record and activity type). Now we also display cells with
other "color" if there exist for that record and that type an activity
corresponding to the filtered "color".

[IMP] crm, {test_}mail: allows to filter record having "done" activity

To allow to search records having "done" activity (archived) assigned to
someone, we have modified the search method of the field activity_user_id 
(in the activity mixin) so that it matches also responsible of archived activities.
This field is mainly used when coming from the systray to display user
activities including "done" ones but could be also used in custom filters.

We have also modified the default domain when fetching record for the activity
view so that it matches records having at least one activity, archived or not.
This prevents to load more records than necessary to display the activity view.

We also adapt the test accordingly.

[IMP] mail, various: simplify the systray

The system tray has been simplified:
- only one clickable area per activity group
- by default, a click on an activity group open the list view, except for some
model (ex.: journal entries, project, task, documents where the activity view
is opened). This can be defined per model using the _systray_view attribute on
the model.
- the clock icon providing access to the activity view has been removed. As it
is the only action, we also remove "actions" returned by systray_get_activities
in res_users.
- the KPIs are no longer clickable

The tests have been updated (for example, because we open the list view instead
of the kanban view) and one removed "activity menu widget: activity view icon"
as we only have one clickable area now and no more icons.

[IMP] various: add activity view for different model

We add the activity view for:
- event.event
- maintenance.request
- mrp.production, mrp.unbuild
- repair.order
- stock.picking

[IMP] various: improve custom activity views

The following summary view of the record in the activity views have been
improved:
- project_task: task state added
- project_project: project manager added
- event_event: responsible and the dates (date_begin_located and
date_end_located) added
- account_move: total amount, customer and state added
- sale_order: total amount and the state added
- purchase_order: total amount and the state added
- crm_lead: customer and stage added
- hr_applicant (hr_recruitment): recruiter added
- survey_survey: responsible added
- maintenance.request: added equipment and responsible
- stock.picking: added scheduled_date
- repair.order: responsible, schedule_date and product_id added
- mrp.production: added responsible
- hr_leave (hr_holidays): status is added and default deadline modified see
below

Ensures that "Schedule activity" is in one line by adding a colspan.

hr_leave activity: When a time off approval activity is created as a
consequence of the creation of a hr.leave, the deadline of the activity is set
to the date_from of the hr.leave minus activity type delay_count (default 15)
except if it leads to a date anterior to today. In that case it is set to
today. That way, in the activity view, the time off approval activities are
more or less sorted by related hr.leave date_from and the cell date is an
indication of when the time off is planned.

Technical note: on most activity view, the activity record was not occupying
all the horizontal space. To solve that problem the css has been modified and
the max-width (200) that was imposed on the sub div has been removed. And as
it was impossible to impose a max-width for a flex div (which is the common
case for the activity record), the max width is imposed on each text that might
be too long using the class o_text_block. That class has been modified to
impose a max-width. That's why that class has been added in most view.
Alternativly, we could have modified the activity compiler to add that class
when the attribute full was set (not done because not sure of the consequence).

[REM] project_todo: remove add todo from the systray

In order to simplify the systray, we remove the action "Add a To-do" from the
activity menu in the systray.

Technical note: We remove the following code in project_todo:
- static/src/web/activity: which add the "Add a To-do" in activity menu in the
systray with its logic
- controllers: which add specific controller for the "Add a To-do" in the
systray
- static/tests: which test the todo from the activity menu
- static/tests/helpers: that supports the test that we have removed (ex.:
simulate the removed controller)

Task-3300854

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/138135).

std-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -571,53 +584,129 @@ def action_open_document(self):
-            'activity_res_ids': sorted(res_id_to_deadline, key=lambda item: res_id_to_deadline[item]),
-            'grouped_activities': activity_data,
+            'activity_res_ids': (  # record with ongoing activities followed by the ones with all completed activities
+                    sorted(res_id_to_deadline, key=lambda item: res_id_to_deadline[item]) +
+                    sorted({k: v for k, v in res_id_to_date_done.items() if k not in res_id_to_deadline},
+                           key=lambda item: res_id_to_date_done[item], reverse=True)
```
is it really necessary ? can't we do just (test are green with that)
```py
sorted(res_id_to_deadline, key=lambda item: res_id_to_deadline[item])
```

(because if the dateline is set, datedone will always be set)

std-odoo replied:
and so the variable `res_id_to_date_done` is now useless

std-odoo reviewed (commented):
@pyduf Hi :) 

Good work as always :) I already post the comments I have (mainly about the python)

Cheers

std-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -571,53 +584,129 @@ def action_open_document(self):
+                                    for attachment_id in activity['attachment_ids']]
+            attachments_info = {}
+            if activity_attachments:
+                last_attachment = max(activity_attachments, key=lambda a: a['create_date'])
```
```suggestion
                last_attachment = max(activity_attachments, key=lambda a: (a['create_date'], a['id']))

std-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -81,6 +81,10 @@ def _default_activity_type_for_model(self, model):
     automated = fields.Boolean(
         'Automated activity', readonly=True,
         help='Indicates this activity has been created automatically and not by any user.')
+    attachment_ids = fields.Many2many(
```
for which activities type can we have many attachment ?
(`upload_file` upload only one attachment, and all the code I find about `action_feedback` (`markAsDone` in JS), and `action_feedback_schedule_next` only use 1 attachment)

If we use a many2one instead, the code of `get_activity_data` can also be simplified

std-odoo replied:
also, looks like `action_feedback_schedule_next` never use attachment, maybe it can be cleaned ?

std-odoo replied:
note to keep a trace: we will keep a m2m

std-odoo reviewed (commented):
@pyduf Hi :) 

Some more comments, notably about the `attachment_ids` field, I wonder if it can be simplified to use a many2one instead (and simplify the methods in js / python).

Cheers

std-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -128,12 +134,12 @@ def _compute_res_name(self):
             tz = record.user_id.sudo().tz
             date_deadline = record.date_deadline
-            record.state = self._compute_state_from_date(date_deadline, tz)
+            record.state = 'done' if not record.active else self._compute_state_from_date(date_deadline, tz)
```
> the compute automatically set the state to “done” when archiving

but if there's no deadline but it's archived, it won't be marked as done no ?

because of that line
```py
for record in self.filtered(lambda activity: activity.date_deadline):

pyduf commented:
> @pyduf Hi :)
> 
> Some more comments, notably about the `attachment_ids` field, I wonder if it can be simplified to use a many2one instead (and simplify the methods in js / python).
> 
> Cheers

Hello @std-odoo , 

Thanks for the review and all your help IRL. I have updated the code and answered your last comments.

std-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -571,53 +584,160 @@ def action_open_document(self):
+
+        # 1. Retrieve all ongoing and completed activities according to the parameters
+        activity_types = self.env['mail.activity.type'].search([('res_model', 'in', (res_model, False))],
+                                                               order='sequence')
```
just to have something determinist if some have the same sequence

```suggestion
                                                               order='sequence, id')

pyduf replied:
Yes, good idea

std-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -571,53 +584,160 @@ def action_open_document(self):
+            ['activity_type_id', 'res_id', 'date_deadline', 'user_id', 'active'] +
+            (['attachment_ids'] if fetch_done else []),
+        )
+        all_ongoing_activities, all_completed_activities = partition(lambda a: a['active'], all_activities)
```
oh didn't know that function :O I might start to use it :)

std-odoo commented on addons/mail/static/src/views/web/activity/activity_renderer.js:
```diff
@@ -92,21 +109,31 @@ export class ActivityRenderer extends Component {
+            }
         }
 
+        const ongoingActivityCount = types.overdue.value + types.today.value + types.planned.value;
```
the file follow lint rules defined in /web/tooling/eslintrc.json before your change, can you check it's still the case ?

pyduf replied:
Thanks, the linter is now working well in my ide.

std-odoo commented on addons/test_mail/tests/test_mail_activity.py:
```diff
@@ -717,6 +722,232 @@ def test_my_activity_flow_employee(self):
+            login='employee2',
+        )
+
+    def get_grouped_activity_for_compare(self, grouped_activities):
```
can be private

std-odoo commented on addons/test_mail/tests/test_mail_activity.py:
```diff
@@ -717,6 +722,232 @@ def test_my_activity_flow_employee(self):
+            attachment_ids=attachment_1.ids)
+        activity_data = self.env['mail.activity'].get_activity_data('mail.test.activity', None, fetch_done=True)
+        expected_grouped_activity[self.test_record_2.id][self.type_upload.id] = {
+            'attachments': { 'last': {'id': attachment_1.id, 'name': 'Uploaded doc_1'}, 'count': 1 },
```
```suggestion
            'attachments': {'last': {'id': attachment_1.id, 'name': 'Uploaded doc_1'}, 'count': 1},

std-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -571,53 +584,160 @@ def action_open_document(self):
+        :param list domain: record search domain
+        :param int limit: maximum number of records to fetch
+        :param int offset: offset of the first record to fetch
+        :return dict: {'activity_types': list of activity type info as list
```
`fetch_done` is missing before that line
(you can move the explanation above here)

std-odoo reviewed (commented):
@pyduf Hi :) 

Some polyshing comments, otherwise LGTM :) 

Cheers

pyduf commented:
> @pyduf Hi :)
> 
> Some polyshing comments, otherwise LGTM :)
> 
> Cheers

Hello @std-odoo, thanks, I have updated the code and answered your comments.

std-odoo commented on addons/calendar/models/res_users.py:
```diff
@@ -51,8 +52,9 @@ def systray_get_activities(self):
-                'icon': modules.module.get_module_icon(self.env['calendar.event']._original_module),
+                'icon': modules.module.get_module_icon(Model._original_module),
                 'meetings': meetings_lines,
+                "view_type": Model._systray_view if hasattr(Model, '_systray_view') else 'list',
```
```suggestion
                "view_type": getattr(Model, '_systray_view', 'list'),

std-odoo commented on addons/mass_mailing_sms/models/res_users.py:
```diff
@@ -17,6 +17,8 @@ def systray_get_activities(self):
         """
         activities = super(Users, self).systray_get_activities()
+        Model = self.env['mailing.mailing']
+        view_type = Model._systray_view if hasattr(Model, '_systray_view') else 'list'
```
same here

std-odoo reviewed (approved):
It will trigger a ci/security check, but it's more correct

Otherwise LGTM :)

tde-banana-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
+            ['activity_type_id', 'res_id', 'date_deadline', 'user_id', 'active'] +
+            (['attachment_ids'] if fetch_done else []),
+        )
+        all_ongoing_activities, all_completed_activities = partition(lambda a: a['active'], all_activities)
```
You tend to use long variable names (but the naming is correct so that's cool :) ), here the "all" prefix does not seem necessary ? It eases reading as there are less "generic" word to analyze... like
```suggestion
        ongoing_activities, completed_activities = partition(lambda a: a['active'], all_activities)
```

tde-banana-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
+            attachments_by_id = {}
+
+        # 3. Group activities per records and activity type
+        grouped_completed_activities = dict(groupby(all_completed_activities,
+                                                    key=lambda a: (a['res_id'], a['activity_type_id'][0])))
+        grouped_ongoing_activities = dict(groupby(all_ongoing_activities,
+                                                  key=lambda a: (a['res_id'], a['activity_type_id'][0])))
```
Example of simpler variable name that does not impact readong IMO
```suggestion
        grouped_completed = dict(
            groupby(completed_activities, key=lambda a: (a['res_id'], a['activity_type_id'][0]))
        )
        grouped_ongoing = dict(
            groupby(ongoing_activities, key=lambda a: (a['res_id'], a['activity_type_id'][0]))
        )

tde-banana-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
+                                                  key=lambda a: (a['res_id'], a['activity_type_id'][0])))
+
+        # 4. Filter out unreadable records
+        all_res_id_activity_type_id = grouped_ongoing_activities.keys() | grouped_completed_activities.keys()
+        if not is_filtered:
+            res_ids = list({res_id for res_id, __ in all_res_id_activity_type_id})
+            res_ids_set = set(Model.search([('id', 'in', res_ids)])._ids)
+            all_res_id_activity_type_id = [(res_id, activity_type_id)
+                                           for res_id, activity_type_id in all_res_id_activity_type_id
+                                           if res_id in res_ids_set]
```
Try to keep a simple naming and coding style when possible, as you go through a lot of intermediate variables and type convert that can make code a bit hard to follow, like
```suggestion
        res_ids_types_tuples = grouped_ongoing.keys() | grouped_completed.keys()
        if not is_filtered:
            filtered = Model.search([('id', 'in', [r[0] for r in res_ids_types_tuples)]).ids
            res_ids_types_tuples = list(filter(lambda x: x[0] in filtered, res_ids_types_tuples))
``

tde-banana-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
+                                   for a in ongoing_activities)),
+                    **({'done': len(completed_activities)} if completed_activities else {}),
+                },
+                'ids': [a['id'] for a in itertools.chain(ongoing_activities, completed_activities)],
```
Try to use recordsets (actually I am not sure what is inside those variables), to make code easier to read, it should be
```suggestion
                'ids': (ongoing_activities | completed_activities).ids,
``

tde-banana-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
             'activity_types': activity_type_infos,
-            'activity_res_ids': sorted(res_id_to_deadline, key=lambda item: res_id_to_deadline[item]),
-            'grouped_activities': activity_data,
+            'activity_res_ids': (  # record with ongoing activities followed by the ones with all completed activities
```
I ... have no clue what it holds ^^ .

tde-banana-odoo commented on addons/mail/models/mail_activity.py:
```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
 
     @api.model
-    def get_activity_data(self, res_model, domain, limit=None, offset=0):
+    def get_activity_data(self, res_model, domain, limit=None, offset=0, fetch_done=False):
```
After reading 'get_activity_data' twice, have to say it could benefits from some simplification / rewriting ... added various comments before this one, globally: try to use short (and meaningful) variables names, use recordsets, and... well... not sure but it seems to have a lot of dict(tuple(list(set)))) and we don't really know what lies inside the final variables ^^ . Would say this needs to be smoother.

tde-banana-odoo commented on addons/test_mail/tests/test_mail_activity.py:
```diff
@@ -717,6 +722,232 @@ def test_my_activity_flow_employee(self):
 
 
+@tests.tagged('mail_activity')
+class TestMailActivity(TestActivityCommon):
```
Same here, according to me: simplification required ! Avoid making "too complete" tests, because I have to say I would probably not be able to update this test if something fails of changes. According to me, you should keep tests simple, two IDs, one type, one variation with "keep done"; avoid multi employee (not sure what is adds). Don't re-do tests, having one past and one future activity is sufficien tfor example as we consider state is already tested, ...

We globally need one test ensuring we have most of the code covered; here it seems to me we are adding a lot of scheduling and activities to be sure to cover everything, but with a lot of redundant code...

KISS (keep it super simple) !

tde-banana-odoo commented on addons/event/views/event_event_views.xml:
```diff
@@ -158,6 +158,26 @@
         </field>
     </record>
 
+    <record id="project_project_view_activity" model="ir.ui.view">
```
project_project_view_activity -> event_event_view_activity :) (copy pasta :D ? )

tde-banana-odoo reviewed (commented):
Went a bit through main python code, have to say I would gladly see some simplification / clarity in code ^^ added some comments with examples on how to simplify the code: shorter variables, less redirection / intermediate variables, ... and maybe check how to simplify globally the computation, as at the end of the main data method I have no clue what it holds :D . It is Friday afternoon so probably a bit tired, but I think it requires a small simplification / cleaning pass.

Cheers !

pyduf commented:
> Went a bit through main python code, have to say I would gladly see some simplification / clarity in code ^^ added some comments with examples on how to simplify the code: shorter variables, less redirection / intermediate variables, ... and maybe check how to simplify globally the computation, as at the end of the main data method I have no clue what it holds :D . It is Friday afternoon so probably a bit tired, but I think it requires a small simplification / cleaning pass.
> 
> Cheers !

Hello @tde-banana-odoo , 

Thanks for the review and the suggestion to simplify the code. I have updated the code and answered your comments.
(Note I will update the code today to correct the style issue and solve some minor problems noticed during the testing)

tde-banana-odoo commented on addons/mail/static/src/views/web/activity/activity_cell.js:
```diff
@@ -25,21 +35,22 @@ export class ActivityCell extends Component {
-                year: "numeric",
-            });
-        }
+    get closestDateFormatted() {
```
I remember we discussed about it and I asked some simplification, but after a new year it would be weird actually ... maybe revert the change ?

tde-banana-odoo commented on addons/mail/static/src/views/web/activity/activity_renderer.js:
```diff
@@ -118,18 +150,26 @@ export class ActivityRenderer extends Component {
     }
 
+    isTypeDisplayDone(typeId) {
+        return this.props.activityTypes.find((a) => a[0] === typeId)[3];
```
mmmh ? Don't we have a better structure to make it clearer than accessing [3] ?

tde-banana-odoo commented on addons/mail/static/src/views/web/activity/activity_renderer.js:
```diff
@@ -118,18 +150,26 @@ export class ActivityRenderer extends Component {
-                .filter(([, resIds]) => typeId in resIds && resIds[typeId].state === name)
-                .map(([key]) => parseInt(key));
+            this.activeFilter.resIds = new Set(
+                Object.entries(this.props.groupedActivities)
+                    .filter(
+                        ([, resIds]) => typeId in resIds && name in resIds[typeId].count_by_state
+                    )
+                    .map(([key]) => parseInt(key))
```
Would reduce a bit line spread here to better understand the code, like
```suggestion
                Object.entries(this.props.groupedActivities)
                    .filter(([, resIds]) => typeId in resIds && name in resIds[typeId].count_by_state)
                    .map(([key]) => parseInt(key))
```

tde-banana-odoo commented on addons/mail/static/tests/helpers/mock_server/models/mail_activity.js:
```diff
@@ -118,57 +126,167 @@ patch(MockServer.prototype, {
+        ]);
+        // Remove domain term used to filter record having "done" activities (not understood by the getRecords mock)
+        domain = Domain.removeDomainLeaves(new Domain(domain ?? []).toList(), [
+            "activity_ids.active",
+        ]).toList();
+        const allRecords = this.getRecords(res_model, domain ?? []);
         const records = limit ? allRecords.slice(offset, offset + limit) : allRecords;
+        const activityDomain = [["res_model", "=", res_model]];
```
Globally, same naming proposals apply here compared to python, if possible to quickly go through and simplify a bit :) (like allOngoingActivities -> ongoing, groupedCompletedActivities -> groupCompleted, ...)

tde-banana-odoo commented on addons/mail/static/tests/helpers/mock_server/models/mail_activity.js:
```diff
@@ -118,57 +126,167 @@ patch(MockServer.prototype, {
+                ).map((a) => a.user_id)
+            );
+            const closestDate = ongoingActivities.length ? dateDeadline : dateDone;
+            const activityAttachments = completedActivities
+                .map((activity) => activity.attachment_ids)
+                .flat()
+                .map((attachmentId) => attachmentsById[attachmentId]);
```
Example of updated naming to gain some lines and ease reading
```suggestion
            const activityAttach = completed.map((act) => act.attachment_ids).flat()
                .map((attachId) => attachmentsById[attachId]);
```

tde-banana-odoo commented on addons/mail/static/tests/helpers/mock_server/models/mail_activity.js:
```diff
@@ -184,12 +302,16 @@ patch(MockServer.prototype, {
-                records.map((x) => x.id),
-                (id) => deserializeDate(resIdToDeadline[id])
-            ),
+            activity_res_ids: sortBy(Object.keys(resIdToDeadline), (item) => resIdToDeadline[item])
```
:eyes:

tde-banana-odoo commented on addons/test_mail/static/tests/systray_activity_menu_tests.js:
```diff
@@ -76,104 +82,64 @@ QUnit.test("activity menu widget: activity menu with 2 models", async (assert) =
-    await click(".o_menu_systray .dropdown-toggle:has(i[aria-label='Activities'])");
-    await click(":nth-child(1 of .o-mail-ActivityGroup) button", { text: "1 Today" });
+    assert.ok(
+        selectContaining(document, ".o-mail-ActivityMenu .o-mail-ActivityGroup", [
+            "res.partner",
+            "0 Late",
+        ])
```
We don't need to take that much vertical space, especially for tests ^^
```suggestion
        selectContaining(document, ".o-mail-ActivityMenu .o-mail-ActivityGroup", ["res.partner", "0 Late"])
```

tde-banana-odoo reviewed (commented):
Thanks for the update :) ! Added a new batch of comments, went through the whole diff :) it is indeed already clearer, I added some suggestions to try to continue the readability improvement. Added some other suggestion also (notably a 'date_done' field, I think it could be better than rewriting on deadline).

Cheers !

pyduf commented:
> Thanks for the update :) ! Added a new batch of comments, went through the whole diff :) it is indeed already clearer, I added some suggestions to try to continue the readability improvement. Added some other suggestion also (notably a 'date_done' field, I think it could be better than rewriting on deadline).
> 
> Cheers !

Hello @tde-banana-odoo , 

Thanks for the review. As suggested, I have added the date_done field in the mail activity model. It is indeed cleaner.
(the code is updated but the runbot is not yet fully build)

tde-banana-odoo commented on addons/repair/views/repair_views.xml:
```diff
@@ -3,6 +3,25 @@
     <data>
 
+    <record id="repair_order_view_activity" model="ir.ui.view">
+        <field name="name">repair.order.activity</field>
```
:socks: such reviewer
```suggestion
        <field name="name">repair.order.view.activity</field>
```

tde-banana-odoo reviewed (commented):
Thanks ! Some small remaining remarks on my side then we can launch :)

tde-banana-odoo commented:
@robodoo r+ rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message.

pyduf commented:
Hello @tde-banana-odoo and @std-odoo, 

Thanks for all your support and helpful advice on this task.
